pkg/autoupgrade/daemon.go

```diff
@@ -0,0 +1,379 @@
+			}
+		}
+	}
+
+	// This loop iterates over d.appKeysToNextCheck (which represents all the apps that have autoUpgrade turned on) and does the following:
+	// 1. If the app no longer exists in the general apps map, remove it, because it must no longer exist
+	// 2. Checks to see if the app has a specific interval set. If it does, and that isn't the interval used on the last run, recalculate the "next check" time
+	// 3. If the app no longer has autoUpgrade turned on, remove it from appKeytsToNextCheck. It must have been turned off since last run
+	for k, nextCheck := range d.appKeysToNextCheck {
+		app, ok := apps[k]
+		if !ok {
+			delete(d.appKeysToNextCheck, k)
+			continue
+		}
+
+		if _, ok := Mode(app.Spec); ok {
+			// Note: if we're using the default interval, nextCheck.appSpecificInterval is ""
+			if nextCheck.appSpecificInterval != app.Spec.AutoUpgradeInterval {
+				next, interval, err := calcNextCheck(defaultNextCheck, app)
+				if err != nil {
+					logrus.Errorf("Problem calculating next check time for app %v: %v", app.Name, err)
+					continue
+				}
+				d.appKeysToNextCheck[k] = nextCheckDetails{time: next, appSpecificInterval: interval}
+			}
+		} else {
+			// App no longer has auto-upgrade enabled. Remove it
+			delete(d.appKeysToNextCheck, k)
+		}
+	}
+
+	// d.appKeysToNextCheck is no fully up-to-date. This loop iterates over it and compares each app's nextCheck time
+	// to the current time. If it nextCheck is before Now, then it is time to check the app.
+	// The refresh map is used to group apps by their image. Checking for new versions of an image is relatively expensive
+	// because it has to go out to an external registry. So, if many apps are using the same image, we just want to pull
+	// the tags for that image once.  The namespace is in the key because pull credentials are namespace specific.
+	refresh := map[imageAndNamespaceKey][]kclient.ObjectKey{}
+	now := time.Now()
+	for appKey, nextCheck := range d.appKeysToNextCheck {
+		app, ok := apps[appKey]
+		if !ok {
+			continue
+		}
+
+		// If next check time is before now, app is due for a check
+		if nextCheck.time.Before(now) {
+			img := app.Status.AppImage.Name
+			if img == "" {
+				img = removeTagPattern(app.Spec.Image)
+			}
+			imageKey := imageAndNamespaceKey{image: img, namespace: app.Namespace}
+			appKeys := refresh[imageKey]
+			refresh[imageKey] = append(appKeys, appKey)
+
+		}
+	}
+
+	// This loop iterates over the refresh map and looks for new versions of image being used for each app.
+	// If it determines a newer version of an image is available for an app, it will update the app with that information
+	// which will trigger the appInstance handlers to pick up the change and deploy the new version of the app
+	for imageKey, appsForImage := range refresh {
+		current, tags, err := pull.ListTags(ctx, d.client, imageKey.namespace, imageKey.image)
+		if err != nil {
+			logrus.Errorf("Problem listing tags for image %v: %v", imageKey.image, err)
```

Comment:
are you sure you don't want to `continue` here and the errors below?

Reply:
I am not using `continue` because an image can either come from a real remote registry or from the local registry, even if it looks like `docker.io/foo/bar:1`, it can still just be a local image. In that scenario, pull.ListTags will actually return an error, but there may be legit local images that match it still.

I think I should take this down to info level and reword it.